### DISCUSSION TITLE
[V5] Creating a Service Provider

### DISCUSSION THREAD

**Discussion by remidosol** [adonisjs/core#1831] [Help] [answered]
Hello there, I'm wondering that how can I create **service provider**. 

There is no page that tell me about providers in V5 docs and I'm trying to create a provider that will be use for ***CDN(AWS S3) transactions***.

I've created a provider but I don't know how can I bind my service with its **Config** file to IoC container. I've created a config file that named as "services.ts" in config folder but when I look at to **other** config files (e.g. database.ts etc.), there is a **type** for this typed config file in **ioc:Adonis/../..** (e.g. DatabaseConfig type etc.). 

If I define these types in config file that I've created, does it will be effective? Otherwise where should I define these types for using this config file (services.ts) in the provider that I've create?

Actually this not the main problem, this was just an example. Is there a template for my problem? Thanks already.

**Comment by Banashek** [comment]
I don't think Service Provider docs exist for v5 yet.

My process for figuring out how they worked included:

1. Look at [the v4 documentation](https://adonisjs.com/docs/4.1/service-providers)
2. Look at [the fold source](https://github.com/adonisjs/fold) to get an idea of how things wire up, specifically how [it really just boils down to calling `register` for each provider, then calling `boot` for each provider](https://github.com/adonisjs/fold/blob/fffb76bb1466dc090a0890efb8821e518d204085/src/Registrar/index.ts#L131) 
3. Looking at how some of the official providers work

Confusingly enough, the [v4 service provider documentation](https://adonisjs.com/docs/4.1/service-providers) states "The register method is used to register bindings, and you should never try to use any other binding inside this method", which is slightly confusing since some of the official providers [explicitly call `use` with other service providers](https://github.com/adonisjs/lucid/blob/develop/providers/DatabaseProvider.ts#L24), however this rule might not apply when referencing core adonisjs providers. I have to look into this more to see if those are loaded before everything else.

Since it seems like the algorithm the application container follows is "call register on everything" -> "call boot on everything", I'm _guessing_ that the general rule is:

1. Call as much as you can in `register`, but don't rely on any of the behavior of another provider's `boot` method (or any other services outside of `core` or registered later than this service).
2. Call anything that relies on other services' `register` results in `boot`.

To answer your concrete question, I'd look to the [Official Redis Provider](https://github.com/adonisjs/redis/blob/develop/providers/RedisProvider.ts#L64) to see how it uses the `Config` to configure it's service.

**Comment by steschwa** [comment, marked as answer]
I've recently faced a similar problem (creating a service provider for S3 file management).

1. Run `node ace make:provider ImageServiceProvider` to create and register your Service Provider
2. Create the service (i've created a seperate folder `/services` inside the project root). Let's call it `S3Service` for now with the following stub:

```typescript
import { MultipartFileContract } from "@ioc:Adonis/Core/BodyParser"

export class S3Service {
    public async upload(file: MultipartFileContract): Promise<void> {
        // Upload the file to your S3 provider ...
    }

    public async download(file: string): Promise<Readable> {
        // Stream file from S3 ...
    }
}
```

3. Register the `S3Service` class to get resolved by your Service Provider:

**providers/ImageServiceProvider.ts**
```typescript
export default class ImageServiceProvider {
    public register() {
        this.container.singleton("MyCompany/ImageService", () => {
            const S3Service = require("../services/S3Service").S3Service // Maybe this can be improved (with import(...)?) 🤷‍♂️
            return new S3Service()
        })
    }
}
```

4. Inform Typescript about your custom IOC binding:

**contracts/ImageService.d.ts**
```typescript
declare module "@ioc:MyCompany/ImageService" {
    import { S3Service } from "services/S3Service"
    const imageService: S3Service
    export default imageService
}
```

5. Import the IOC binding into your Controller etc.

**app/Controllers/Http/ExampleController**
```typescript
import { HttpContextContract } from "@ioc:Adonis/Core/HttpContext"
import ImageService from "@ioc:MyCompany/ImageService"

export default class ExampleController {
    public async store({ request }: HttpContextContract) {
        // Validation ...
       
        if (validated.image?.tmpPath) {
            await ImageService.upload(validated.image) // 👈 with Typescript support
        }
    }
}
```

Because there isn't an official documentation yet, this solution may not be perfect.

You may also see [here](https://github.com/adonisjs/core/discussions/1545) and [here](https://github.com/adonisjs/core/discussions/1420).

**Reply by remidosol** [reply]
I have to accept one mime type that is "multipart/form-data" from HTTP Request. So, I guess I'm gonna get the image file**S** using ***request.files()*** method. 

In your example's "app/Controllers/Http/ExampleController" part, there is a key that is named as **tmpPath**, what does this **if**'s conditional statement control?

**Reply by steschwa** [reply]
If you don't want to validate your files (which i strongly discourage) you can use `request.files()` ([Typed in **adonisjs/bodyparser**](https://github.com/adonisjs/bodyparser/blob/c1c9cc2c660a5cc37cc7282e281fa6c29925f1b0/adonis-typings/request.ts))

With `validated.image?.tmpPath` i check if the file is available on my local disk before uploading it to S3.

**Reply by remidosol** [reply]
> If you don't want to validate your files (which i strongly discourage) you can use `request.files()` ([Typed in **adonisjs/bodyparser**](https://github.com/adonisjs/bodyparser/blob/c1c9cc2c660a5cc37cc7282e281fa6c29925f1b0/adonis-typings/request.ts))
> 
> With `validated.image?.tmpPath` i check if the file is available on my local disk before uploading it to S3.

**First**: Maybe we can validate these files like below?
**In a validation file:**
```typescript
    images: schema.array([
      rules.minLength(1),
      rules.maxLength(3)
    ]).members(
      schema.file({
        size: '1mb',
        extnames: ['.jpg']
      })
    )
```
And images's type looks like that:
![Ow](https://i.hizliresim.com/89jk7R.jpg)

**Second**: For uploading file/files to **S3**, is the file/files ***must be available on a local storage?*** Is is a necessity?

**Reply by steschwa** [reply]
1) Should work
2) The user sends the file to your API (and is temporary stored on your API server). I thought if the upload failed for some reason, your server cannot redirect (upload) it to the S3. That's why i've used `if(validated.image?.tmpPath) { ... }`
